### DISCUSSION TITLE
Vault Credential Manager and AppRole difficulty

### DISCUSSION THREAD

**Discussion by Jazun713** [concourse/concourse#5641] [General]
So I'm having a bit of difficulty getting concourse to use vault as a credential manager with an approle. 

I have successfully configured it to use a client token.  Currently, I'm setting everything as environment variables in a docker container using the official concourse image concourse/concourse:6.0.0

I can pull a secret using the following config:
`    environment:
      - CONCOURSE_POSTGRES_HOST=concourse-db
      - CONCOURSE_POSTGRES_USER=concourse_user
      - CONCOURSE_POSTGRES_PASSWORD=concourse_pass
      - CONCOURSE_POSTGRES_DATABASE=concourse
      - CONCOURSE_EXTERNAL_URL
      - CONCOURSE_ADD_LOCAL_USER=admin:admin
      - CONCOURSE_MAIN_TEAM_LOCAL_USER=admin
      - CONCOURSE_VAULT_URL=https://example.com:8200
      - CONCOURSE_VAULT_CA_CERT=/opt/certs/vault_ca.crt
      - CONCOURSE_VAULT_SHARED_PATH=/concourse/shared
      - CONCOURSE_SECRET_RETRY_ATTEMPTS=5   # how many times to try
      - CONCOURSE_SECRET_RETRY_INTERVAL=10s # how long to wait between attempts
      - CONCOURSE_VAULT_CLIENT_TOKEN=xxxxxxxx
      - CONCOURSE_VAULT_INSECURE_SKIP_VERIFY=true`

But if I exchange `CONCOURSE_VAULT_CLIENT_TOKEN=xxxxx` with:
`     - CONCOURSE_VAULT_AUTH_BACKEND="approle"
      - CONCOURSE_VAULT_AUTH_PARAM="role_id:xxxxx,secret_id:xxxxxx"`

I get the following error message in my pipeline:
`var lookup 'git_password': Error making API request.

URL: GET https://example.com:8200/v1/concourse/main/example-pipeline/git_password
Code: 400. Errors:

* missing client token`

Any thoughts as to what's causing this?  I can log into my vault instance using the role_id and secret_id and pull this same secret using the vault CLI and the API from both my docker host (laptop) and inside the container.

**Comment by Jazun713** [comment]
github completely mutilated my code block.

Here's a gist of my docker-compose.yml with the environment variables
https://gist.github.com/Jazun713/0e94337b058f75bf7f132bc5bf07a4d8

**Comment by Jazun713** [comment]
Has anyone seen this?

**Comment by vito** [comment]
Just to confirm: have you followed the [`approle` docs](https://concourse-ci.org/vault-credential-manager.html#vault-approle-auth)? It's possible they're out of date though.

Do you see any logs from the `web` node about failed login attempts?

**Reply by Jazun713** [reply]
I followed those docs for sure.  I didn't see anything in the web node logs about failed login attempts I can take another look today and confirm.

**Comment by Jazun713** [comment]
@vito 
Here's a sample of the log below.
There's definitely some login failures but they're all saying the token is missing.  It's hard to read but my config is as above.  I'm using CONCOURSE_VAULT_AUTH_BACKEND and CONCOURSE_VAULT_AUTH_PARAM as environment variables.   This is a docker container in this instance.  my compose file is in the gist linked above.

> concourse_1     | {"timestamp":"2020-06-04T15:20:05.204768600Z","level":"info","source":"atc","message":"atc.cmd.start","data":{"session":"1"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:05.346119600Z","level":"info","source":"atc","message":"atc.credential-manager.configured credentials manager","data":{"name":"vault","session":"6"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:05.871665100Z","level":"error","source":"atc","message":"atc.credential-manager.login.failed","data":{"error":"Error making API request.\n\nURL: PUT https://vault.example.com:8200/v1/auth/%22approle%22/login\nCode: 400. Errors:\n\n* missing client token","name":"vault","session":"6.1"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:07.205772100Z","level":"error","source":"atc","message":"atc.credential-manager.login.failed","data":{"error":"Error making API request.\n\nURL: PUT https://vault.example.com:8200/v1/auth/%22approle%22/login\nCode: 400. Errors:\n\n* missing client token","name":"vault","session":"6.2"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:09.186338400Z","level":"error","source":"atc","message":"atc.credential-manager.login.failed","data":{"error":"Error making API request.\n\nURL: PUT https://vault.example.com:8200/v1/auth/%22approle%22/login\nCode: 400. Errors:\n\n* missing client token","name":"vault","session":"6.3"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.487764800Z","level":"info","source":"atc","message":"atc.cmd.finish","data":{"duration":141500,"session":"1"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.491631900Z","level":"info","source":"baggageclaim","message":"baggageclaim.using-driver","data":{"driver":"overlay"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.493523600Z","level":"info","source":"tsa","message":"tsa.listening","data":{}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.493849100Z","level":"info","source":"atc","message":"atc.scheduler-interval-runner.start","data":{"interval":"10s","session":"20"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.494548300Z","level":"info","source":"atc","message":"atc.lidar.start","data":{"session":"17"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.494612600Z","level":"info","source":"worker","message":"worker.garden.dns-proxy.started","data":{"session":"1.2"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.494643200Z","level":"info","source":"atc","message":"atc.lidar.start","data":{"session":"17"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.495143200Z","level":"info","source":"atc","message":"atc.listening","data":{"debug":"127.0.0.1:8079","http":"0.0.0.0:8080"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.495267300Z","level":"info","source":"baggageclaim","message":"baggageclaim.listening","data":{"addr":"127.0.0.1:7788"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.516790000Z","level":"info","source":"tsa","message":"tsa.connection.channel.command.forwarded-tcpip","data":{"bind-addr":"0.0.0.0:7777","bound-port":37723,"command":"forward-worker","remote":"127.0.0.1:57130","session":"1.4.1"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.516838900Z","level":"info","source":"tsa","message":"tsa.connection.channel.command.forwarded-tcpip","data":{"bind-addr":"0.0.0.0:7788","bound-port":38725,"command":"forward-worker","remote":"127.0.0.1:57130","session":"1.4.1"}}
concourse_1     | {"timestamp":"2020-06-04T15:20:10.516880000Z","level":"info","source":"tsa","message":"tsa.connection.channel.command.start","data":{"command":"forward-worker","remote":"127.0.0.1:57130","session":"1.4.1"}}

**Reply by Jazun713** [reply]
Do i need to generate a login token with this app role first and include that in the config to get this to work?

Wouldn't that just use the client token indefinitely and not need the approle?

**Comment by Jazun713** [comment]
I'm still looking for help on this.

**Comment by gowrisankar22** [comment]
Did you check this https://github.com/concourse/concourse/discussions/5738 ?

**Reply by Jazun713** [reply]
@gowrisankar22 I just read through it.  I'm able to interpolate variables if I use a vault client token but I get an error if I try to use just an app role.  Do I need both? Doesn't seem like I do.

**Reply by gowrisankar22** [reply]
@Jazun713 Nope if you are using approle remove the token.  Just with these two variables, it should work. Also depends on how you create an approle.
]
```
CONCOURSE_VAULT_AUTH_BACKEND="approle"
CONCOURSE_VAULT_AUTH_PARAM="role_id:5f3420cd-3c66-2eff-8bcc-0e8e258a7d18,secret_id:f7ec2ac8-ad07-026a-3e1c-4c9781423155"
```

To verify approle is working, can you try to create a token using the approle roleid and secretid and try to read the secret. if that is working, we can check the next thing.
```
vault write auth/approle/login \
    role_id=5f3420cd-3c66-2eff-8bcc-0e8e258a7d18 \
    secret_id=f7ec2ac8-ad07-026a-3e1c-4c9781423155
```

**Reply by Jazun713** [reply]
This is working

**Reply by gowrisankar22** [reply]
I didn't find `CONCOURSE_VAULT_PATH_PREFIX=/concourse` in your gist link. Also, can you comment out `CONCOURSE_VAULT_CA_CERT` and try?
